pkg/bootiso/bootiso.go

```diff
@@ -0,0 +1,116 @@
+package bootiso
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"github.com/u-root/u-root/pkg/boot"
+	"github.com/u-root/u-root/pkg/boot/kexec"
+	"github.com/u-root/u-root/pkg/boot/syslinux"
+	"github.com/u-root/u-root/pkg/mount"
+	"github.com/u-root/u-root/pkg/mount/loop"
+	"golang.org/x/sys/unix"
+)
+
+// ParseConfigFromISO mounts an iso file to a
+// temp dir to get the config options
+func ParseConfigFromISO(isoPath string) ([]boot.OSImage, error) {
+	tmp, err := ioutil.TempDir("", "mnt")
+	if err != nil {
+		return nil, fmt.Errorf("Error creating mount dir: %v", err)
+	}
+
+	loopdev, err := loop.New(isoPath, "iso9660", "")
+	if err != nil {
+		return nil, fmt.Errorf("Error creating loop device: %v", err)
+	}
+
+	mp, err := loopdev.Mount(tmp, unix.MS_RDONLY|unix.MS_NOATIME)
+	if err != nil {
+		return nil, fmt.Errorf("Error mounting loop device: %v", err)
+	}
+	defer mp.Unmount(0)
+
+	configOpts, err := syslinux.ParseLocalConfig(context.Background(), tmp)
+	if err != nil {
+		return nil, fmt.Errorf("Error parsing config: %v", err)
+	}
+
+	return configOpts, nil
+}
+
+// BootFromPmem copies the ISO to pmem0 and boots
+// given the syslinux configuration with the provided label
+func BootFromPmem(isoPath string, configLabel string) error {
+	pmem, err := os.OpenFile("/dev/pmem0", os.O_APPEND|os.O_WRONLY, 0600)
+	if err != nil {
+		return fmt.Errorf("Error opening persistent memory device: %v", err)
+	}
+
+	iso, err := os.Open(isoPath)
+	if err != nil {
+		return fmt.Errorf("Error opening ISO: %v", err)
+	}
+	defer iso.Close()
+
+	if _, err := io.Copy(pmem, iso); err != nil {
+		return fmt.Errorf("Error copying from ISO to pmem: %v", err)
+	}
+	if err = pmem.Close(); err != nil {
+		return fmt.Errorf("Error closing persistent memory device: %v", err)
+	}
+
+	tmp, err := ioutil.TempDir("", "mnt")
+	if err != nil {
+		return fmt.Errorf("Error creating temp directory: %v", err)
+	}
+
+	if _, err := mount.Mount("/dev/pmem0", tmp, "iso9660", "", unix.MS_RDONLY|unix.MS_NOATIME); err != nil {
+		return fmt.Errorf("Error mounting pmem0 to temp directory: %v", err)
+	}
+
+	configOpts, err := syslinux.ParseLocalConfig(context.Background(), tmp)
```

Comment:
The main issue I was trying to solve was figuring out the parameters to pass to `kexec()`. The  `syslinux.ParseLocalConfig()` function seemed like a good way to do this. If you mount the ISO and then call this function, it parses the ISO's syslinux.cfg file and returns an array of LinuxImage objects that each contain:
- Path to the kernel
- Path to the initramfs
- List of kernel parameters

There were 2 alternatives I was considering when I thought about how to call the function.
1. Mount the ISO temporarily and retrieve the LinuxImage objects. Later the user can decide whether to boot the ISO from pmem or directly from the USB stick. Because we mount temporarily and then remount the ISO at the boot step, we need to retrieve the updated LinuxImage objects with the final paths to the kernel and initramfs. _This was the version implemented here._
2. With the ISO in the location it will be booted from (either pmem or the USB stick), mount the ISO and retrieve the LinuxImage objects. When the user selects the desired LinuxImage, pass it directly to `kexec()`

A final alternative would be to hardcode the boot parameters in a lookup table like they do in the original webboot.

Reply:
Yup the reasoning makes sense, I was just wondering if there was a way we could reuse some of the logic in ParseConfigFromISO here. But I see the difference now, thanks!